### ISSUE TITLE
Let cargo test and cargo commands with --examples ignore examples that are not for the current target

### ISSUE THREAD

**Issue by najamelan** [rust-lang/cargo#7233]
**Describe the problem you are trying to solve**
Currently `cargo test --target wasm32-unkown-unkown` will try to compile all examples. This is a problem if there is code that uses dev-dependencies that are not available on this target, like networking crates.

**Describe the solution you'd like**
This already works for files in the `tests` directory, but not for examples:
I would like to see an example ignored when it has: `#![ cfg(not( target_arch = "wasm32" )) ]` at the top of the file. 

Currently this throws an error like:
```shell
error[E0601]: `main` function not found in crate `basic`
  |
  = note: consider adding a `main` function to `examples/basic.rs`
```

The only workaround I have found is to put a cfg attribute on each toplevel statement and to add a dummy main tagged with a cfg for wasm. This is problematic if you have a bunch of free standing functions and type definitions, especially since examples are a form of documentation. 

Ideally also cargo would have a nice error message if a user tries to run an example on the wrong target. That is, if there is a positive cfg limiting the example to a single target, telling them to run it with `cargo run --example basic --target wasm32-unknown-unknown` and if there is a negative cfg, telling them: This example cannot run on target xxx.

**Comment by najamelan** [comment]
ps: I haven't tested what the score is with features, but I imagine it would be handy to in/exclude examples depending on whether a feature is turned on or not.

**Comment by najamelan** [comment]
It turns out cargo test doesn't know how to run library test (not integration tests) in wasm:
```
cargo test --target wasm32-unknown-unknown                                           426ms  zo 11 aug 2019 23:53:25 CEST ~ yesnet
   Compiling futures_cbor_codec v0.1.0 (/data/doc/code/futures_cbor_codec)
    Finished dev [unoptimized + debuginfo] target(s) in 0.13s
     Running target/wasm32-unknown-unknown/debug/deps/futures_cbor_codec-e7b6100345028991.wasm
/data/doc/code/futures_cbor_codec/target/wasm32-unknown-unknown/debug/deps/futures_cbor_codec-e7b6100345028991.wasm: /data/doc/code/futures_cbor_codec/target/wasm32-unknown-unknown/debug/deps/futures_cbor_codec-e7b6100345028991.wasm: cannot execute binary file
error: test failed, to rerun pass '--lib'
```
Adding `--lib` didn't really solve it. I have used wam-pack to test integration tests that actually run in the browser quite successfully, but this is a general purpose crate that doesn't really use anything wasm specific.

**Comment by najamelan** [comment]
I have a similar problem with a crate that compiles on stable, but has examples that require nightly. Now cargo test can no longer be run on stable.

**Comment by najamelan** [comment]
I made some progress without modifying cargo:

- I tried `package.autoexamples = false` + `[[target."cfg(not( target_arch = \"wasm32\" ))".example]]` but that failed. It seems that the target selection is specific to dependencies?
- I managed a workaround with `package.autoexamples = false`, a `notwasm` feature and using `required-features = notwasm` in the example section. The downside of this is that users can only run examples with: `cargo run --example basic --features notwasm`, which is unfortunate and also one needs to list all examples in `Cargo.toml`. I suppose this technique could also work with a `nightly` feature to allow tests to run in `stable`.
